lib/std/zon/parse.zig

```diff
@@ -0,0 +1,2832 @@
+        node: Zoir.Node.Index,
+    ) !T {
+        switch (node.get(self.zoir)) {
+            .string_literal => return try self.parseString(T, node),
+            .array_literal => |nodes| return try self.parseSlice(T, options, nodes),
+            .empty_literal => return try self.parseSlice(T, options, .{ .start = node, .len = 0 }),
+            else => return self.failExpectedContainer(T, node),
+        }
+    }
+
+    fn parseString(
+        self: *@This(),
+        comptime T: type,
+        node: Zoir.Node.Index,
+    ) !T {
+        const ast_node = node.getAstNode(self.zoir);
+        const pointer = @typeInfo(T).pointer;
+        var size_hint = ZonGen.strLitSizeHint(self.ast, ast_node);
+        if (pointer.sentinel_ptr != null) size_hint += 1;
+
+        var buf: std.ArrayListUnmanaged(u8) = try .initCapacity(self.gpa, size_hint);
+        defer buf.deinit(self.gpa);
+        switch (try ZonGen.parseStrLit(self.ast, ast_node, buf.writer(self.gpa))) {
+            .success => {},
+            .failure => |err| {
+                const token = self.ast.nodes.items(.main_token)[ast_node];
+                const raw_string = self.ast.tokenSlice(token);
+                return self.failTokenFmt(token, @intCast(err.offset()), "{s}", .{err.fmt(raw_string)});
+            },
+        }
+
+        if (pointer.child != u8 or
+            pointer.size != .slice or
+            !pointer.is_const or
+            (pointer.sentinel_ptr != null and @as(*const u8, @ptrCast(pointer.sentinel_ptr)).* != 0) or
+            pointer.alignment != 1)
+        {
+            return self.failExpectedContainer(T, node);
+        }
+
+        if (pointer.sentinel_ptr != null) {
+            return try buf.toOwnedSliceSentinel(self.gpa, 0);
+        } else {
+            return try buf.toOwnedSlice(self.gpa);
+        }
+    }
+
+    fn parseSlice(
+        self: *@This(),
+        comptime T: type,
+        comptime options: Options,
+        nodes: Zoir.Node.Index.Range,
+    ) !T {
+        const pointer = @typeInfo(T).pointer;
+
+        // Make sure we're working with a slice
+        switch (pointer.size) {
+            .slice => {},
+            .one, .many, .c => @compileError(@typeName(T) ++ ": non slice pointers not supported"),
+        }
+
+        // Allocate the slice
+        const sentinel = if (pointer.sentinel_ptr) |s| @as(*const pointer.child, @ptrCast(s)).* else null;
+        const slice = try self.gpa.allocWithOptions(
+            pointer.child,
+            nodes.len,
+            pointer.alignment,
+            sentinel,
+        );
+        errdefer self.gpa.free(slice);
+
+        // Parse the elements and return the slice
+        for (0..nodes.len) |i| {
+            errdefer if (options.free_on_error) {
+                for (slice[0..i]) |item| {
+                    free(self.gpa, item);
+                }
+            };
+            slice[i] = try self.parseExpr(pointer.child, options, nodes.at(@intCast(i)));
+        }
```

Comment:
```suggestion
        for (slice, 0..) |*elem, i| {
            errdefer if (options.free_on_error) {
                for (slice[0..i]) |item| {
                    free(self.gpa, item);
                }
            };
            elem.* = try self.parseExpr(pointer.child, options, nodes.at(@intCast(i)));
        }
```
This makes the data flow a little clearer (we can immediately see that each loop iteration is going to initialize `slice[i]`).